src/PooledArrays.jl

```diff
@@ -30,13 +30,13 @@ end
 
 mutable struct PooledArray{T, R<:Integer, N, RA} <: AbstractArray{T, N}
     refs::RA
-    pool::Vector{T}
-    invpool::Dict{T,R}
+    pool::Vector{Union{Missing,T}}
```

Comment:
I kind of think we should just keep this `T`; and let `Union{Missing, T}` be provided as the `T`. That way if you don't have `missing` values, you're not penalized. It seems like to so `simillar_missing`, we could provide a function that just re-wraps the `refs`, `pool`, `invpool`, etc in a new `PooledArray` struct w/ `Union{T, Missing}`, right?

But maybe you're right that it would be too expensive to have to change `pool::Vector{T}` and `invpool::Dict{T, R}` to support `missing` as well in the `similar_missing` operation.

Reply:
I am trying to avoid having to pay the following cost:
```
julia> y = Dict{Int, Int}(1:10^8 .=> 1:10^8);

julia> @time convert(Dict{Union{Int, Missing}, Int}, y);
  3.684057 seconds (253.19 k allocations: 4.514 GiB, 6.35% gc time, 2.88% compilation time)
```
(it would also reduce the memory burden of making this operation)

With this PR it would be avoided.
My original preference was different, i.e. to never allow `Missing` in the `refpool` and `invrefpool` but have `0` to be a ref that is interpreted as `missing` without making a lookup, but @nalimilan thought it would be overly complex. That is why I have made a draft here to discuss the design.

This would have a consequence similar to what we have in CategoricalArrays.jl:
```
julia> x = categorical(["a"])
1-element CategoricalArray{String,1,UInt32}:
 "a"

julia> resize!(x, 2)
2-element CategoricalArray{String,1,UInt32}:
    "a"
 #undef

julia> convert(CategoricalArray{Union{String,Missing},1,UInt32}, x)
2-element CategoricalArray{Union{Missing, String},1,UInt32}:
 "a"
 missing
```
(so magically `#undef` becomes `missing` which is probably undesirable)

There is no rush to make a decision here. Let us think and decide what is best.

Reply:
Yeah neither solution is simple. Maybe it would be possible to make `convert` faster for dicts when the target eltype is a supertype of the source eltype?

Reply:
AFAICT this has been an open issue in Julia Base for a long time. Who do you think is best to ask?

Reply:
No idea. I guess we (you :-p) would have to come up with at least a draft PR to get people's attention. But this may be quite easy to do. `convert` calls this constructor when the origin type is the same as the destination:
https://github.com/JuliaLang/julia/blob/be6887fa9ffa3d08315c93f1551c41b44dd3d720/base/dict.jl#L92-L95
I wonder whether what's needed is just to call `convert` on the keys and value when the types don't match. What do you think?

Assuming that works, do you think it would be an acceptable replacement for this PR or would it still not be sufficient?

Reply:
I do not think this would work. The code of the problem is the following code:
```
julia> x = [1]
1-element Vector{Int64}:
 1

julia> y = convert(Vector{Union{Int, Missing}}, x)
1-element Vector{Union{Missing, Int64}}:
 1
```
If I understand things correctly `x` and` y` have a different memory layout and some allocation must happen. What I would want to have is that it would be possible that two `PooledVectors` - one allowing `Missing` and the other not allowing it could share the same pool internally (until Copy-On-Write happens - of course).

Reply:
Ah yes, that's https://github.com/JuliaLang/julia/issues/26681. But at least copying a vector should be much cheaper than copying the dict (which currently rehashes all entries)?

Reply:
Yes, but I would want to investigate the option to fully avoid copying (as it seems it should be possible). Out of the approaches we discussed (the one in this PR or the one in CategoricalArrays.jl) - which one would you prefer?

Reply:
I'm leaning towards the approach in this PR; I'm close to having my CSV.jl refactor working which assumes this representation and that seems to simplify some things, at least on the manual construction side.

Reply:
TBH I prefer the solution where we don't have to change anything in PooledArrays, and instead improve `convert` in Julia to avoid copies. :-)

Reply:
But then we loose COW benefits. However, since we are not fully convinced what is best I would park this PR for now, so that @quinnj can finish CSV.jl update without this functionality (which is more pressing). We can go back to this PR and decide what to do later (especially that if we change the internals probably 2.0 release would be needed).

And maybe in the mean time what @nalimilan proposes to do in Julia Base would be implemented (I cannot do it as I do not know how to do it unfortunately as I know too little about internal design related to https://github.com/JuliaLang/julia/issues/26681)

Reply:
I'm afraid the best person to implement no-copy `convert` for `Array{Union{T, Missing}}` is also @quinnj! :-D

Reply:
@quinnj - if you do not have time for this (which I assume is the case) - can you please point me to the part of the sources that would have to be changed? (I assume it is somewhere in the C codes)
Thank you!

Reply:
All the relevant code is in https://github.com/JuliaLang/julia/blob/master/src/array.c. In the [jl_array_typetagdata](https://github.com/JuliaLang/julia/blob/master/src/array.c#L52) function, it returns the "selector bytes" for an isbits-union array. [This](https://github.com/JuliaLang/julia/blob/master/src/array.c#L104) is where we lengthen the requested array size to account for selectory bytes. We also have to ensure the selector bytes are always [initialized to zero](https://github.com/JuliaLang/julia/blob/master/src/array.c#L140), since otherwise you could get selector bytes pointing to something not just garbage, but totally wrong. 

So the operation we'd need is `convert(::Type{Vector{Union{T, S}}}, x::Vector{T}) where {T, S}`, right? And we're ok if the two arrays potentially share data? So it seems we would add a check on the julia-side for if `T` is `isbitstype` and `S` is `isbitstype`, and if so, call a new array.c routine.

This is where it gets a little hairy though, because we have [`array_resize_buffer`](https://github.com/JuliaLang/julia/blob/master/src/array.c#L673), but that takes an existing array to resize it. We'd need to use some of the same logic in that routine and make a new `jl_convert_to_isbitsunion` method that:
* asserts the input array eltype is `isbitstype`
* asserts new desired `Union` eltype is also `isbitstype`
* checks that the existing array data isn't `a->flags.isshared` (means we can't resize it; like an mmap buffer)
* then probably called the same gc method as [this line](https://github.com/JuliaLang/julia/blob/master/src/array.c#L696) to resize existing array data to new size (which is just old array size + length(old_array))
* then I think we could just call [`jl_ptr_to_array_1d`](https://github.com/JuliaLang/julia/blob/master/src/array.c#L327) with our realloc'd pointer? that will take care of making the new array and using the realloc'd data pointer

Reply:
Yeah - I was afraid it would not be that simple. What we need is only `convert(::Type{Vector{Union{T, Missing}}}, x::Vector{T}) where {T}` but `T` may be any type (also non-bits, e.g. `String`). So I guess it would require separate paths for `T` being bits type and being non-bits type.

However, this in general shows the following problem. Assume that we have `x` and `y` sharing the same data. Where eltype of `x` is `T` and of `y` is `Union{Missing, T}`. What will happen to `x` if we do `y[1] = missing`? My intuition that this operation is super problematic - right? (i.e. would lead to an undefined state of `x[1]`)

So the conclusion is that these arrays may not share data. Therefore we have an easier task of adding an internal constructor in PooledArrays.jl for `Dict` type that instead of:
```
function Dict{K,V}(kv) where V where K
    h = Dict{K,V}()
    haslength(kv) && sizehint!(h, Int(length(kv))::Int)
    for (k,v) in kv
        h[k] = v
    end
    return h
end
```
would copy an internal structure of source `Dict` and only do promotion of eltype of `keys` field in the source dict.

So we should have a function like:
```
function widen_dict_missing(d::Dict{K,V}) where {K, V}
    @assert !(Missing <: K) # this is handled otherwise in PooledArrays.jl
    K′ = Union{K, Missing}
    return Dict{K′,V}(copy(d.slots), Vector{K′}(d.keys), copy(d.vals),
                      d.ndel, d.count, d.age, d.idxfloor, d.maxprobe)
end
```
and the performance would be:
```
julia> d = Dict(1:10^7 .=> 1:10^7);

julia> @time Dict{Union{Int,Missing}, Int}(d);
  0.282199 seconds (7 allocations: 288.001 MiB)

julia> @time Dict{Union{Int,Missing}, Int}(d);
  0.301356 seconds (7 allocations: 288.001 MiB)

julia> @time Dict{Union{Int,Missing}, Int}(d);
  0.313754 seconds (7 allocations: 288.001 MiB)

julia> @time widen_dict_missing(d);
  0.080447 seconds (7 allocations: 288.000 MiB)

julia> @time widen_dict_missing(d);
  0.078461 seconds (7 allocations: 288.000 MiB)

julia> @time widen_dict_missing(d);
  0.085039 seconds (7 allocations: 288.000 MiB)
```
the benefit would be ~3x speedup.